### ISSUE TITLE
Clarifying usage with custom react apps

### ISSUE THREAD

**Issue by oyeanuj** [interactivethings/catalog#310]
Hi @herrstucki, thank you for putting this library out there - excited to use it. However, I was a little confused about how to set it up, and ran into some issues.

I was using a clone of the popular `react-universally` starter kit to start off. Like most starter kits, it has its own custom webpack setup. So, putting together the instructions in the changelog and the documentation, I did the following:

1. Installed the package and created the catalog. Worked fine, added the commands to package.json and I could access the catalog on `localhost:4000`. All good here (except maybe that it adds 3.0 beta-6 version instead of the release v3)
```sh
yarn add catalog
yarn create catalog
```

2. I added the library components via pages:
```jsx
//index.js
const pages = [
  {
    path: '/',
    title: 'Welcome',
    content: pageLoader(() => import('./WELCOME.md')),
  },
  {
    path: '/elements',
    title: 'Elements',
    content: pageLoader(() => import('./ElementPage.js')),
  },
];

...

//ElementPage.js
import React from 'react';
import ReactDOM from 'react-dom';
import { Page, ReactSpecimen, ColorSpecimen } from 'catalog';
import { Logo, Menu } from '../shared/components/DemoApp/Header'; //<------

export default () => (
  <Page>
    <h2>My Buttons</h2>
    
    <ReactSpecimen span={3}>
      <Logo />
      <Header />
    </ReactSpecimen>

    <ColorSpecimen
      name="Red"
      value="#FF5500"
      span={2}
    />

  </Page>
);
```

At that point, I get the following error:
```js
./shared/components/DemoApp/Header/index.js
Module parse failed: /Users/anuj/Github/react-universally/shared/components/DemoApp/Header/index.js Unexpected token (18:4)
You may need an appropriate loader to handle this file type.
| function Header() {
|   return (
|     <HeaderWrapper>
|       <Logo />
|       <h1>React, Universally</h1>
 @ ./catalog/ElementPage.js 17:14-60
 @ ./catalog/index.js
 @ multi ./node_modules/catalog/dist/cli/config/polyfills.js ./catalog/index.js
```

I wasn't sure how to proceed at this point. I added the `catalog/babel` to my `.babelrc` but I am not even sure if it is being used by catalog?

If I try `yarn catalog-build`, I get the same error as above plus the errors below:
```sh
static/1.589a5ece.chunk.js from UglifyJs
Unexpected token: punc ()) [./catalog/ElementPage.js:6,15][static/1.589a5ece.chunk.js:29,19]

static/catalog.f474a272.js from UglifyJs
Unexpected token: punc ()) [./catalog/index.js:9,24][static/catalog.f474a272.js:19569,37]
```

Without the custom react component, it works well!

---

FWIW: Here's a repo for you to quickly reproduce the issue: https://github.com/oyeanuj/react-universally/tree/dev-with-catalog

Any ideas?

Thank you!

**Comment by wereHamster** [comment]
The problem is that the webpack config which catalog-start generates internally only processes .js files in the `src/` and `catalog/` folders. JavaScript files in other folders are processed with the default loader which presumably doesn't understand jsx.

The generated webpack config is derived from create-react-app (CRA). If your project uses a folder structure that is different from CRA, you'll have to use your own webpack config.

Since your project already supports multiple entry points, I suggest that you define another bundle and set its entry file to `catalog/index.js`. The following changes were enough so that catalog would start up when running `npm run develop`:

 - client bundle: set `srcEntryFile` to `'./catalog/index.js'`, add `'./catalog'` to `srcPaths`.
 - in catalog/index.js, change the DOM element into which the catalog is rendered, from `'catalog'` to `'app'`.

Loading `.md` files doesn't work yet, but it's just a matter of adding the correct loader to the webpack config. See https://docs.catalog.style/guides/webpack-babel.

**Comment by oyeanuj** [comment]
@wereHamster Thank you for the quick response! 

1. [I tried that](https://github.com/oyeanuj/react-universally/commit/91cee75610ed808f135bb871e27646504767dc53) but still seeing the loader errors. Here is the commit: https://github.com/oyeanuj/react-universally/commit/91cee75610ed808f135bb871e27646504767dc53

2. With the above, I wasn't sure where the bundle should be outputted and if it would still be accessible at `localhost:4000` like before?

3. Is it possible to just configure the path `catalog-start` processes as a parameter? That might be simpler and would then work regardless of the host project?

**Comment by wereHamster** [comment]
I see two loader errors: one because you still haven't added a loader for the markdown files, therefore loading of `WELCOME.md` fails. And the second error is because webpack tries to load the logo.png with the javascript loader. I have no idea why that is.

What I meant is to run either your normal application, or catalog. And have a way to switch to one or the other from the commandline. It doesn't make sense to build both bundles, you can't run both at the same time anyways.

The path in the webpack config far from the only option you'd have to change. You use a heavily customised webpack config which you'd have to replicate for catalog. And catalog doesn't currently allow users to modify its webpack config. In my experience, if you use anything but a stock CRA or next.js setup, it's far easier to treat catalog as a library and use it as a plain react component from within your existing application.

**Comment by oyeanuj** [comment]
@wereHamster Firstly, I really appreciate you taking out the time to look over my problem.  Just generally speaking, what are the steps that you'd expect are needed when dealing with custom webpack configs, and build setups?  I feel like maybe I am missing the understanding of what does catalog need to be able to run properly.

```js
1. Add Catalog and `create catalog`.
2. ???
3. Profit!
```

Below is my report back:

> I see two loader errors: one because you still haven't added a loader for the markdown files, therefore loading of WELCOME.md fails. And the second error is because webpack tries to load the logo.png with the javascript loader. I have no idea why that is.

So, I tried removing any reference to markdown or image files and I still see the loader error failing at basic HTML 😞 Here is the commit for reference: https://github.com/oyeanuj/react-universally/commit/45bfa04a9b143af0a39298bb2214773956bb3e96

```jsx
Module parse failed: /Users/anuj/Github/react-universally/shared/components/DemoApp/Header/Logo/index.js Unexpected token (6:9)
You may need an appropriate loader to handle this file type.
| function Logo() {
|   // return <img src={logo} alt="Logo" style={{ width: '100px' }} />;
|   return <h1>This is the logo!</h1>;
| }
|
 @ ./catalog/ElementPage.js 17:12-63
 @ ./catalog/index.js
 @ multi ./node_modules/catalog/dist/cli/config/polyfills.js ./node_modules/react-dev-utils/webpackHotDevClient.js ./catalog/index.js
```

Thoughts on why `catalog` is not able to import the component to render?

---
> What I meant is to run either your normal application, or catalog. And have a way to switch to one or the other from the commandline. It doesn't make sense to build both bundles, you can't run both at the same time anyways.

Following your other tip, I tried to create a bundle to build and run catalog but then I get the following error:
```
ERROR in ./catalog/index.html
Module parse failed: /Users/anuj/Github/react-universally/catalog/index.html Unexpected token (1:0)
You may need an appropriate loader to handle this file type.
| <!doctype html>
| <html>
| <head>
 @ multi regenerator-runtime/runtime react-hot-loader/patch webpack-hot-middleware/client?reload=true&path=http://0.0.0.0:7331/__webpack_hmr ./catalog/index.html
```

---
> The path in the webpack config far from the only option you'd have to change. You use a heavily customised webpack config which you'd have to replicate for catalog. And catalog doesn't currently allow users to modify its webpack config. 

I can see that. Maybe in the upcoming versions, allowing users to specify file paths, like Storybook does, could be possible?
```js
const req = require.context('../shared/components', true, /.stories.js$/);

function loadStories() {
  req.keys().forEach(filename => req(filename));
}
```
---
Again, thank you for your help! 😄

**Comment by wereHamster** [comment]
The first problem of not being able to load Logo.js is again because the file is outside of `catalog/` or `src/`. If you rename `shared/` to `src/` then `yarn run catalog-start` will work.

`yarn run magic` fails because the entry file (`srcEntryFile`) must be a JavaScript file. I believe that's a restriction from webpack itself. If you change `srcEntryFile: 'catalog/index.html'` to `srcEntryFile: 'catalog/index.js'` then you'll at least be able to run `yarn run magic` and it will produce a bunch of files in `build/catalog/`, though without an corresponding index.html file and therefore the output is pretty useless.

As I said, it's not just the paths. Users need the ability to modify the whole webpack config (to add custom loaders and whatnot). What you need is a solution for #304.

**Comment by oyeanuj** [comment]
@wereHamster Thanks for responding to my questions! 

On the happy side, I was able to make things work when the path was `/src`! 

On the not so happy side, introducing a bundle seemed more complex than I anticipated - especially not knowing how to configure the right entry points like you mentioned. 

If you do at some point end up creating an example setup in the future, I'd love to re-attempt.. until then, I guess I'll wait for #304 to migrate from `storybook` to `catalog`.

**Comment by herrstucki** [comment]
@oyeanuj I released v3.1.0 which adds support for a config file and also always Babel-transpiles from the app's root folder instead of just from `src/`. Let me know if there are any issues!

**Comment by krystof-k** [comment]
Hey there, I have a related question (I guess). I'm trying to setup Catalog with externally loaded create-react-app based React components from another repo, added as a dependency. Do you have any clue how to do that? When running Catalog directly in the create-react-app repo everything works fine, but we would like to separate the catalog to another repo. Thanks!
